### PR TITLE
feat: ocsf audit logging

### PR DESCRIPTION
**Pull request by durg78**
Added OCSF (version 1.2.0) audit log format utilizing github.com/valllabh/ocsf-schema-golang.

The dependency for golangci-lint was updated from 1.54 to 1.59.1.  Lint test was crashing, and updating it resolved the issue.

ocsf-schema-golang required an update from go 1.20 to 1.22.  Consequently, reflect.StringHeader (used in internal/corazawaf/rule.go) was deprecated in go 1.21.  This was causing lint check failures.  I have created an exception for now, but this will need to be addressed.

### CODE REVIEW

File: coraza.conf-recommended
Comment (jcchavezs) [inline coraza.conf-recommended:235]: I don't think we want to recommend this by default as this feels like a breaking change.

File: .golangci.yml
Comment (jcchavezs) [inline .golangci.yml:25]: Do we truly need this?

File: examples/http-server/go.mod
Comment (jcchavezs) [inline examples/http-server/go.mod:3]: while this is great it break our policy of 2 latest versions supported.

File: experimental/plugins/plugintypes/auditlog.go
Comment (jcchavezs) [inline experimental/plugins/plugintypes/auditlog.go:34]: Could you add a comment on what is this? I know others don't have comments but I think since this is OCSF specific, worth checking.

File: experimental/plugins/plugintypes/auditlog.go
Comment (jcchavezs) [inline experimental/plugins/plugintypes/auditlog.go:65]: Same for comments.

File: internal/corazawaf/waf.go
Comment (jcchavezs) [inline internal/corazawaf/waf.go:123]: Audit log is pluggable hence users might define their own, I don't think we can support a strict type for this.

File: internal/corazawaf/transaction.go
Comment (jcchavezs) [inline internal/corazawaf/transaction.go:1377]: I truly doubt the internal transaction has to know about all these fields. I wonder if this could happen in the internal/auditlog#Transaction type.

**Comment by jcchavezs** [comment]
I think this is great. Being a existing format honestly feels this is better than our JSON format or the modsec one which I am not 100% sure fits in modern tooling. My only request here is that we should avoid changing the transaction to support a format. Audit log formats are pluggable and we should be able to either do everything with what we have or extend what we have to support new use cases but that does not include changing the internal API.

File: coraza.conf-recommended
Comment (fzipi) [reply coraza.conf-recommended:235]: Agreed. We should keep the native as default for now.

File: .golangci.yml
Comment (durg78) [reply .golangci.yml:25]: reflect.StringHeader was deprecated in go 1.21,  this was causing the lint check to fail.  I wasn't sure how best to handle that other than adding an exception for now.

File: experimental/plugins/plugintypes/auditlog.go
Comment (durg78) [reply experimental/plugins/plugintypes/auditlog.go:34]: Can do.

File: .golangci.yml
Comment (jptosso) [reply .golangci.yml:25]: We can also use inline linter nochecks

File: examples/http-server/go.mod
Comment (durg78) [reply examples/http-server/go.mod:3]: Sorry, wasn't aware of the policy.  I can make this work without this change.

File: .golangci.yml
Comment (M4tteoP) [reply .golangci.yml:25]: I think the lint version is going to be related to the Go version. So once you downgrade back to `go 1.20`, The current linter (v1.54) should work back again

**Comment by fzipi** [comment]
Hey @durg78 ! We update go to >= 1.21 today, if you want to rebase.

**Comment by durg78** [comment]
> Hey @durg78 ! We updated go to >= 1.21 today, if you want to rebase.

I just now saw this.  I'm working on the rebase now.

File: .golangci.yml
Comment (durg78) [reply .golangci.yml]: I was able to remove this.  PR #1145 now includes a one liner to exclude the static check for the use of reflect.StringHeader

File: coraza.conf-recommended
Comment (durg78) [reply coraza.conf-recommended]: I've reverted the default back to "Native".

File: examples/http-server/go.mod
Comment (durg78) [reply examples/http-server/go.mod]: resolved via PR #1145

File: experimental/plugins/plugintypes/auditlog.go
Comment (durg78) [reply experimental/plugins/plugintypes/auditlog.go]: Comments added

File: experimental/plugins/plugintypes/auditlog.go
Comment (durg78) [reply experimental/plugins/plugintypes/auditlog.go]: Comments added

File: internal/corazawaf/transaction.go
Comment (durg78) [reply internal/corazawaf/transaction.go]: I was able to move this into formats_ocsf.go.

File: go.work
Comment (fzipi) [inline go.work:3]: Is this whitespace change neede?

File: coraza.conf-recommended
Comment (fzipi) [inline coraza.conf-recommended:165]: There is also a `jsonlegacy` somehow....

**Review by fzipi — COMMENTED** [review]
Probably we could add a little more tests coverage, but this is looking very good now.

File: go.work
Comment (durg78) [reply go.work]: This was inadvertent and has been removed.

**Comment by M4tteoP** [comment]
Would you mind rebasing the PR to fix the conflict and bringing in TinyGo's latest version (https://github.com/corazawaf/coraza/pull/1148 has been merged a few days ago)? I'm curious to see if the Tiny go test now works good

**Comment by fzipi** [comment]
Checks are passing now! 🎉

File: internal/seclang/directives.go
Comment (M4tteoP) [inline internal/seclang/directives.go:651]: For completeness' sake, let's also add here the `JsonLegacy`. These are the comment lines from where the website pages are generated. 
```suggestion
// Syntax: SecAuditLogFormat JSON|JsonLegacy|Native|OCSF
```

**Review by fzipi — APPROVED** [review]
LGTM. Waiting on @M4tteoP's approval also.

File: internal/corazawaf/transaction.go
Comment (M4tteoP) [inline internal/corazawaf/transaction.go:1392]: `Args()`, `Length()`, and `UID()` have been added to the `AuditLogTransactionRequest` interface. The OCSF Formatter then reads these values, but here I see that only `Args_` is populated. Also, in the test, I don't see their value being checked. As far as I get this, Length() and UID() will always just return the default (0 and "") values. Should also these values be populated here?

File: experimental/plugins/plugintypes/auditlog.go
Comment (M4tteoP) [inline experimental/plugins/plugintypes/auditlog.go:68]: What is the request UID in the generic context of a request? How does it differ from the transaction ID that we already have?
I see that you are using the ID in the Metadata of WebResourcesActivity, and then using the new UID() in the HttpRequest, so there might be a different meaning that I'm missing. Would be great to have a comment line here to explain that.

**Review by M4tteoP — COMMENTED** [review]
Just wish to understand a couple of fields that have been added to the `AuditLogTransactionRequest interface` before approving this

File: experimental/plugins/plugintypes/auditlog.go
Comment (durg78) [reply experimental/plugins/plugintypes/auditlog.go:68]: Good catch!
Length was intended to capture the size of the request in bytes.  It should be logging transaction.variables.fullRequestLength here.

I struggled with deciding if transaction.id was better logged in WebResourcesActivity.Metadata.uid (The logging system-assigned unique identifier of an event instance) or WebResourcesActivity.HTTPRequest.uid (The unique identifier of the http request).  Now that you've got me looking at this again.  I think this actually does belong in the http request and not the metadata.

It looks like I had started to build this out but never completed it. I'll tidy this up and add some tests for them.

Comment (M4tteoP) [reply experimental/plugins/plugintypes/auditlog.go:68]: Awesome, thanks!

Comment (durg78) [reply experimental/plugins/plugintypes/auditlog.go:68]: After looking through the code it appears `transaction.variables.FullRequestLength` is never populated with data.  If we're not currently capturing this information, would it be preferred that I simply remove the Length field from `AuditLogTransactionRequest` or simply comment it out with a note about future use?

Comment (fzipi) [reply experimental/plugins/plugintypes/auditlog.go:68]: Hmmm... I would say use it (it will be 0 for now), but let's create an issue and followup PR with that data populated ;)

**Review by M4tteoP — APPROVED** [review]
LGTM, Thanks for all the dedication around this PR @durg78!

**Comment by fzipi** [comment]
Time to merge!

**Comment by jcchavezs** [comment]
Amazing work!

**Pull request merged**
